### ISSUE TITLE
Feed Items without content

### ISSUE THREAD

**Issue by aschilling** [fguillot/picoFeed#218]
Hi Bernhard,

I really like PicoFeed and use it via miniflux nearly everyday. Congratulations for the great work. Unfortunately I found that PicoFeed does not parse the content of some feed items from: http://acemaxx-analytics-dispinar.blogspot.com/feeds/posts/default

In particular the following feed items are parsed without content:

"Reverse Repo in Begleitung zum Leitzins" from 27.11.2015
"Lockere Geldpolitik der EZB und EUR" from 26.11.2015

I really appreciate if you could have a look at these items.

Best Regards

**Comment by BernhardPosselt** [comment]
Hm, I did not find anything unusual so far, @fguillot @mkresin 

PS: @fguillot started the project and Miniflux, I'm working on https://github.com/owncloud/news/ ;)

**Comment by aschilling** [comment]
Sorry for the missunderstanding. Can you reproduce the items with the missing content?

**Comment by BernhardPosselt** [comment]
Yep, can reproduce it in the News app, articles are empty after (including) "Why Minsky Matters"

**Comment by mkresin** [comment]
Yeah, it's a picofeed issue, since the item content is discarded.:

```
$ ./picofeed debug "http://acemaxx-analytics-dispinar.blogspot.com/feeds/posts/default"

Array
(
    [0] => [2015-12-02 09:19:45] PicoFeed\Client\Curl Fetch URL: http://acemaxx-analytics-dispinar.blogspot.com/feeds/posts/default
    [1] => [2015-12-02 09:19:45] PicoFeed\Client\Curl Etag provided: 
    [2] => [2015-12-02 09:19:45] PicoFeed\Client\Curl Last-Modified provided: 
    [3] => [2015-12-02 09:19:46] PicoFeed\Client\Curl cURL total time: 0.352434
    [4] => [2015-12-02 09:19:46] PicoFeed\Client\Curl cURL dns lookup time: 0.028305
    [5] => [2015-12-02 09:19:46] PicoFeed\Client\Curl cURL connect time: 0.057494
    [6] => [2015-12-02 09:19:46] PicoFeed\Client\Curl cURL speed download: 371309
    [7] => [2015-12-02 09:19:46] PicoFeed\Client\Curl cURL effective url: http://acemaxx-analytics-dispinar.blogspot.com/feeds/posts/default
    [8] => [2015-12-02 09:19:46] PicoFeed\Client\HttpHeaders HTTP status code: 200
    [9] => [2015-12-02 09:19:46] PicoFeed\Client\HttpHeaders HTTP header: Content-Encoding => gzip
    [10] => [2015-12-02 09:19:46] PicoFeed\Client\HttpHeaders HTTP header: ETag => W/"7d065a8a-856c-464b-86df-756c91c45a1c"
    [11] => [2015-12-02 09:19:46] PicoFeed\Client\HttpHeaders HTTP header: Date => Wed, 02 Dec 2015 09:19:46 GMT
    [12] => [2015-12-02 09:19:46] PicoFeed\Client\HttpHeaders HTTP header: Content-Type => application/atom+xml; charset=UTF-8
    [13] => [2015-12-02 09:19:46] PicoFeed\Client\HttpHeaders HTTP header: Server => blogger-renderd
    [14] => [2015-12-02 09:19:46] PicoFeed\Client\HttpHeaders HTTP header: Cache-Control => public, must-revalidate, proxy-revalidate, max-age=1
    [15] => [2015-12-02 09:19:46] PicoFeed\Client\HttpHeaders HTTP header: Vary => Accept-Encoding
    [16] => [2015-12-02 09:19:46] PicoFeed\Client\HttpHeaders HTTP header: Expires => Wed, 02 Dec 2015 09:19:47 GMT
    [17] => [2015-12-02 09:19:46] PicoFeed\Client\HttpHeaders HTTP header: X-Content-Type-Options => nosniff
    [18] => [2015-12-02 09:19:46] PicoFeed\Client\HttpHeaders HTTP header: X-XSS-Protection => 1; mode=block
    [19] => [2015-12-02 09:19:46] PicoFeed\Client\HttpHeaders HTTP header: Last-Modified => Wed, 02 Dec 2015 06:15:40 GMT
    [20] => [2015-12-02 09:19:46] PicoFeed\Client\HttpHeaders HTTP header: Content-Length => 130862
    [21] => [2015-12-02 09:19:46] PicoFeed\Client\HttpHeaders HTTP header: X-Frame-Options => SAMEORIGIN
    [22] => [2015-12-02 09:19:46] PicoFeed\Client\HttpHeaders HTTP header: Connection => close
    [23] => [2015-12-02 09:19:46] PicoFeed\Parser\Atom: HTTP Encoding "utf-8" ; XML Encoding "utf-8"
    [24] => [2015-12-02 09:19:46] PicoFeed\Parser\Atom: begin parsing
    [25] => [2015-12-02 09:19:46] PicoFeed\Parser\Atom
Feed::id = tag:blogger.com,1999:blog-686528700264433416
Feed::title = ACEMAXX-ANALYTICS
Feed::feed_url = http://www.blogger.com/feeds/686528700264433416/posts/default
Feed::site_url = http://acemaxx-analytics-dispinar.blogspot.com/
Feed::language = 
Feed::description = 
Feed::logo = 
Feed::date = Wed, 02 Dec 15 07:15:40 +0000
Feed::isRTL() = false
Feed::items = 25 items
----
Item::id = f5872519e929d2d3a3c9966184baeba4eddd41820b58daa02b430174af6d676c
Item::title = Niedrigzinsen und dunkle Ecken der Wirtschaft
Item::url = http://acemaxx-analytics-dispinar.blogspot.com/2015/12/niedrigzinsen-und-dunkle-ecken-der.html
Item::language = 
Item::author = Acemaxx-Analytics
Item::enclosure_url = 
Item::enclosure_type = 
Item::date = Wed, 02 Dec 15 07:15:40 +0000
Item::isRTL() = false
Item::content = 5994 bytes
----
Item::id = 2bc2e1bbb3bb1b2db1c56f9d218be55b3b97c923b0956783fbb5c24396559874
Item::title = Zinsdifferenz: EZB versus SNB
Item::url = http://acemaxx-analytics-dispinar.blogspot.com/2015/12/zinsdifferenz-ezb-versus-snb.html
Item::language = 
Item::author = Acemaxx-Analytics
Item::enclosure_url = 
Item::enclosure_type = 
Item::date = Tue, 01 Dec 15 08:52:49 +0000
Item::isRTL() = false
Item::content = 7787 bytes
----
Item::id = 456f1a3469af94185e2a42d7b5ce4346eaff7aa339cb2b04a13c70d2dd1b0425
Item::title = Angebot-Nachfrage: Austerität und Wachstum
Item::url = http://acemaxx-analytics-dispinar.blogspot.com/2015/11/angebot-nachfrage-austeritat-und.html
Item::language = 
Item::author = Acemaxx-Analytics
Item::enclosure_url = 
Item::enclosure_type = 
Item::date = Mon, 30 Nov 15 08:40:30 +0000
Item::isRTL() = false
Item::content = 5083 bytes
----
Item::id = 60d25ea4114ce383965a4eb786772bc2ad19c940ea7a6372c54c235cedbeae0e
Item::title = Why Minsky Matters
Item::url = http://acemaxx-analytics-dispinar.blogspot.com/2015/11/why-minsky-matters.html
Item::language = 
Item::author = Acemaxx-Analytics
Item::enclosure_url = 
Item::enclosure_type = 
Item::date = Sun, 29 Nov 15 13:03:11 +0000
Item::isRTL() = false
Item::content = 0 bytes
----
Item::id = bd59636be1612202e72349a860940a4b3afd17323fa1b42b4395f8d79e3dc714
Item::title = Reverse Repo in Begleitung von Leitzins
Item::url = http://acemaxx-analytics-dispinar.blogspot.com/2015/11/reverse-repo-in-begleitung-zum-leitzins.html
Item::language = 
Item::author = Acemaxx-Analytics
Item::enclosure_url = 
Item::enclosure_type = 
Item::date = Sat, 28 Nov 15 19:39:10 +0000
Item::isRTL() = false
Item::content = 0 bytes
----
Item::id = ea9224194c84c8fd75727acbe3c43f3d6a6b287c75b8061bce06e4cfe60fc868
Item::title = Lockere Geldpolitik der EZB und EUR
Item::url = http://acemaxx-analytics-dispinar.blogspot.com/2015/11/lockere-geldpolitik-der-ezb-und-eur.html
Item::language = 
Item::author = Acemaxx-Analytics
Item::enclosure_url = 
Item::enclosure_type = 
Item::date = Thu, 26 Nov 15 15:24:04 +0000
Item::isRTL() = false
Item::content = 0 bytes
----
Item::id = 1c5fe16ecf82881a2d7271764c00092d6c5885379fa2f07435490c5826adf071
Item::title = Wahnsinnige Haushaltspolitik und Finanzzyklus
Item::url = http://acemaxx-analytics-dispinar.blogspot.com/2015/11/wahnsinnige-haushaltspolitik-und.html
Item::language = 
Item::author = Acemaxx-Analytics
Item::enclosure_url = 
Item::enclosure_type = 
Item::date = Thu, 26 Nov 15 07:13:12 +0000
Item::isRTL() = false
Item::content = 0 bytes
----
Item::id = d30ff1efe839903063f933ea1e9eca5bab092c81f75ccba1df407e88dbf06459
Item::title = Negative Renditen, Überschussliquidität und Geldpolitik
Item::url = http://acemaxx-analytics-dispinar.blogspot.com/2015/11/negative-renditen-uberschussliquiditat.html
Item::language = 
Item::author = Acemaxx-Analytics
Item::enclosure_url = 
Item::enclosure_type = 
Item::date = Wed, 25 Nov 15 07:20:48 +0000
Item::isRTL() = false
Item::content = 3438 bytes
----
```

The first obvious that I noticed is the usage of awful bloated word html in the content section of the items.

**Comment by mkresin** [comment]
The problem is, that the content of this item can't be (and should not be) converted to XML. The conversation is done by the following call chain:

1. ```Parser::filterItemContent```
2. ```Filter::execute```
3. ```Filter::preFilter```
4. ```Tag::removeBlacklistedTags```
5. ```XmlParser::getDomDocument```
6. ```XmlParser:getSimpleXml```

The bug is triggered by double hyphen within an comment of the following element value (entity decoding done by me for better readability)

```xml
<!-- <m:brkBinSub m:val="--"/> -->
```

within the content section of the item "Reverse Repo in Begleitung von Leitzins"

```xml
<!--[if gte mso 9]><xml> <w:WordDocument>  <w:View>Normal</w:View>  <w:Zoom>0</w:Zoom>  <w:TrackMoves/>  <w:TrackFormatting/>  <w:PunctuationKerning/>  <w:ValidateAgainstSchemas/>  <w:SaveIfXMLInvalid>false</w:SaveIfXMLInvalid>  <w:IgnoreMixedContent>false</w:IgnoreMixedContent>  <w:AlwaysShowPlaceholderText>false</w:AlwaysShowPlaceholderText>  <w:DoNotPromoteQF/>  <w:LidThemeOther>EN-US</w:LidThemeOther>  <w:LidThemeAsian>X-NONE</w:LidThemeAsian>  <w:LidThemeComplexScript>X-NONE</w:LidThemeComplexScript>  <w:Compatibility>   <w:BreakWrappedTables/>   <w:SnapToGridInCell/>   <w:WrapTextWithPunct/>   <w:UseAsianBreakRules/>   <w:DontGrowAutofit/>   <w:SplitPgBreakAndParaMark/>   <w:EnableOpenTypeKerning/>   <w:DontFlipMirrorIndents/>   <w:OverrideTableStyleHps/>  </w:Compatibility>  <m:mathPr>   <m:mathFont m:val="Cambria Math"/>   <m:brkBin m:val="before"/>   <m:brkBinSub m:val="--"/>   <m:smallFrac m:val="off"/>   <m:dispDef/>   <m:lMargin m:val="0"/>   <m:rMargin m:val="0"/>   <m:defJc m:val="centerGroup"/>   <m:wrapIndent m:val="1440"/>   <m:intLim m:val="subSup"/>   <m:naryLim m:val="undOvr"/>  </m:mathPr></w:WordDocument></xml>
<![endif]-->
```

Using double hyphen within comments is fine for html but the [XML standards document](http://www.w3.org/TR/REC-xml/#sec-comments) clearly forbids this:

> [Definition: Comments may appear anywhere in a document outside other markup; in addition, they may appear within the document type declaration at places allowed by the grammar. They are not part of the document's character data; an XML processor may, but need not, make it possible for an application to retrieve the text of comments. For compatibility, the string " -- " (double-hyphen) must not occur within comments.] Parameter entity references must not be recognized within comments.

Long story short: It's a picofeed bug, since picofeed interprets html as xml.

If someone wants to work on this, I've uploaded [my test feed](https://gist.github.com/mkresin/0e3c39d3cf1990f0645a).

**Comment by BernhardPosselt** [comment]
Hm don't we use domdocument's loadHTML method for the HTML stuff?

**Comment by mkresin** [comment]
Only scrapper & favicon related function are using loadHTML.

As far as I can see, the problems start at https://github.com/fguillot/picoFeed/blob/master/lib/PicoFeed/Filter/Html.php#L82, where the html is converted to xml without stripping double-hyphen from comments.

I don't know why we stick to XML within the HTML filter, instead of using loadHTML. Due to lack of time because of other projects I'm working on, I didn't had a detailed look where the (HTML) filter is used else.

**Comment by BernhardPosselt** [comment]
BTW if we go for the loadHTML approach it would be awesome if we could bump the required PHP version to 5.4 and libxml to 2.7.8 since there are very useful libxml headers available, see http://php.net/manual/en/libxml.constants.php
* LIBXML_HTML_NOIMPLIED: doesnt require you to costly remove useless html tags after getting the HTML out of it
* LIBXML_HTML_NODEFDTD: same for doctype

IMHO 5.3 is old enough now, even 5.4 is out of security support and did not receive XML security updates for which we require a full blown lib like ZendXML. It's time to start requiring PHP 5.4 for fixing this issue :)

**Comment by BernhardPosselt** [comment]
Btwi hope that the XML parser that is used to apply the filter rulers itself is not vulnerable to XXE, can you check @mkresin ?

**Comment by mkresin** [comment]
As written in https://github.com/fguillot/picoFeed/issues/218#issuecomment-162027294 ```XmlParser::getSimpleXml``` is used. So the same as for the feed are applied.

But as this is security relevant, I would feel better if you have a look at this as well.

**Comment by BernhardPosselt** [comment]
Yep, getSimpleXml is safe

**Comment by fguillot** [comment]
@BernhardPosselt I'm ok to bump the requirement to PHP 5.4 since the last Debian stable has php 5.6, Ubuntu 14.04 has php 5.5 and Centos 7 has php 5.4. Even if the PHP dev team doesn't support anymore php 5.4, maintainers of Linux distros backport security patches. People who runs older versions of Linux will have to upgrade... so I'm waiting your PR :)

**Comment by BernhardPosselt** [comment]
Ok, will take a look if I have enough time on my hands :)

Solution would be to load the html and then go through each node recursively and apply the filters
